File: part2_oop/src/Ad.cpp
#include "Ad.hpp"

const std::string describeMe_msg = "Advertiser Class methods :\n"
                                   "getTitle() : gives title of the ad.\n"
                                   "setTitle(new_title) : changes title of the ad to new_title.\n"
                                   "getImgUrl() : gives the URL of the ad's image.\n"
                                   "setImgUrl(new_imgUrl) : changes imgUrl of the ad to new_imgUrl.\n"
                                   "getLink() : gives the link of the ad.\n"
                                   "setLink(new_link) : changes link of the ad to new_link.\n"
                                   "getClicks() : gives number of clicks on the ad.\n"
                                   "setAdvertiser(new_advertiser) : changes ad's advertiser to new_advertiser.\n"
                                   "getViews() : gives number of views of the ad.\n"
                                   "incClicks() : increments number of clicks by one.\n"
                                   "incViews() : increments number of views by one.\n"
                                   "describeMe() : describes the methods action.\n";
Ad::Ad(int _id, std::string _title, std::string _imgUrl, std::string _link, Advertiser* _advertiser)
    : BaseAdvertising(_id), title(_title), imgUrl(_imgUrl), link(_link), advertiser(_advertiser) {}  

std::string Ad::getTitle() {
    return title;
}
void Ad::setTitle(std::string _title) {
    title = _title;
}
std::string Ad::getImgUrl() {
    return imgUrl.url;
}
void Ad::setImgUrl(std::string _url) {
    imgUrl = URL(_url);
}
std::string Ad::getLink() {
    return link.url;
}
void Ad::setLink(std::string _link) {
    link = URL(_link);
}
void Ad::setAdvertiser(Advertiser* _advertiser) {
    advertiser = _advertiser;
}
std::string Ad::describeMe() {
    return describeMe_msg;
}
void Ad::incClicks() {
    clicks++;
    advertiser->incClicks();
}
void Ad::incViews() {
    views++;
    advertiser->incViews();
}

File: part3_python/ad.py
from base_advertising import BaseAdvertising
from base_advertising import URL
from advertiser import Advertiser

class Ad(BaseAdvertising):
    def __init__(self, _id, _title, _img_url, _link, _advertiser):
        super().__init__(_id)
        self.title = _title
        self.img_url = URL(_img_url)
        self.link = URL(_link)
        self.advertiser = _advertiser

    def get_title(self):
        return self.title

    def set_title(self, _title):
        self.title = _title

    def get_img_url(self):
        return self.img_url.url

    def set_img_url(self, _url):
        self.img_url = URL(_url)

    def get_link(self):
        return self.link.url

    def set_link(self, _link):
        self.link = URL(_link)

    def set_advertiser(self, _advertiser):
        self.advertiser = _advertiser

    def describe_me(self):
        return "Ad Class methods:\n" \
               "get_title(): gives the title of the ad.\n" \
               "set_title(new_title): changes the title of the ad to new_title.\n" \
               "get_img_url(): gives the URL of the ad's image.\n" \
               "set_img_url(new_img_url): changes img_url of the ad to new_img_url.\n" \
               "get_link(): gives the link of the ad.\n" \
               "set_link(new_link): changes link of the ad to new_link.\n" \
               "get_clicks(): gives the number of clicks on the ad.\n" \
               "set_advertiser(new_advertiser): changes ad's advertiser to new_advertiser.\n" \
               "get_views(): gives the number of views of the ad.\n" \
               "inc_clicks(): increments the number of clicks by one.\n" \
               "inc_views(): increments the number of views by one.\n" \
               "describe_me(): describes the methods action.\n"

    def inc_clicks(self):
        self.clicks += 1
        self.advertiser.inc_clicks()

    def inc_views(self):
        self.views += 1
        self.advertiser.inc_views()

